lib/gitsh/tab_completion/automaton.rb

```diff
@@ -0,0 +1,79 @@
+require 'set'
+
+module Gitsh
+  module TabCompletion
+    class Automaton
+      def initialize(start_state)
+        @start_state = start_state
+      end
+
+      def match(tokens)
+        tokens.inject(start_states) do |current_states, token|
+          current_states.map { |state| state.follow(token) }.inject(Set.new, :|)
+        end
+      end
+
+      def completions(context, token)
+        match(context).
+          flat_map { |state| state.completions(token) }.
+          select { |option| option.start_with?(token) }
```

Comment:
How about moving this to `BaseMatcher`, to make this automaton definition more clear?